File: articles/molecular-biology-flow-of-genetic-information/fortran/molecular_flow_kernel.f90
! Compact molecular information-flow numerical kernel in Fortran.

program molecular_flow_kernel
  implicit none

  real :: m0
  real :: k
  real :: t
  real :: expression
  real :: half_life
  real :: treated
  real :: control
  real :: log2fc
  real :: p_distance
  real :: jc
  real :: mutations
  real :: genomes
  real :: sites
  real :: generations
  real :: mu

  m0 = 100.0
  k = log(4.0) / 4.0
  t = 4.0

  expression = m0 * exp(-k * t)
  half_life = log(2.0) / k

  treated = 160.0
  control = 40.0
  log2fc = log(treated / control) / log(2.0)

  p_distance = 2.0 / 20.0
  jc = -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * p_distance)

  mutations = 12.0
  genomes = 1000.0
  sites = 100000.0
  generations = 1.0
  mu = mutations / (genomes * sites * generations)

  print *, "Transcript at 4h:", expression
  print *, "Half-life:", half_life
  print *, "log2FC:", log2fc
  print *, "Jukes-Cantor distance:", jc
  print *, "Mutation rate:", mu
end program molecular_flow_kernel

File: articles/molecular-biology-flow-of-genetic-information/c/molecular_flow_kernel.c
/*
 * Compact molecular information-flow kernel in C.
 */

#include <math.h>
#include <stdio.h>
#include <string.h>

double transcript_decay(double m0, double k, double t) {
    return m0 * exp(-k * t);
}

double half_life(double k) {
    return log(2.0) / k;
}

double jukes_cantor(double p) {
    if (p >= 0.75) {
        return NAN;
    }

    return -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * p);
}

double gc_fraction(const char *seq) {
    int gc = 0;
    int length = strlen(seq);

    for (int i = 0; i < length; i++) {
        if (seq[i] == 'G' || seq[i] == 'C') {
            gc++;
        }
    }

    return (double)gc / (double)length;
}

double mutation_rate(double mutations, double genomes, double sites, double generations) {
    return mutations / (genomes * sites * generations);
}

int main(void) {
    const char *seq = "ATGCTAGCTAACGGTACCTA";
    double k = log(4.0) / 4.0;
    double log2fc = log(160.0 / 40.0) / log(2.0);
    double p_distance = 2.0 / 20.0;

    printf("transcript_at_4h=%.6f\n", transcript_decay(100.0, k, 4.0));
    printf("half_life_h=%.6f\n", half_life(k));
    printf("gc_fraction=%.6f\n", gc_fraction(seq));
    printf("jukes_cantor=%.6f\n", jukes_cantor(p_distance));
    printf("log2fc=%.6f\n", log2fc);
    printf("mutation_rate=%.12f\n", mutation_rate(12.0, 1000.0, 100000.0, 1.0));

    return 0;
}
